### PR TITLE
KD fix w/ online distillation

### PR DESCRIPTION
**Pull request by winglian**
While top-k before the softmax is more memory efficient, it is mostly incorrect for KD and shouldn't be used. Fixed the collator to ignore non-tensor/list data when we skip data processing and have pretokenized data.

## Summary by CodeRabbit

- **New Features**
  - Introduced advanced knowledge distillation (KD) capabilities, including online teacher support, new KD strategies, and dynamic KD temperature scheduling.
  - Added plugin-based extensibility for training arguments and data collators, enabling custom training workflows.
  - Integrated a memory-efficient, chunked top-K KL-divergence loss for KD with optional hard and soft loss components.
  - Provided new DeepSpeed configuration for Torch compile optimization.

- **Improvements**
  - Enhanced chat template logic and formatting for certain models.
  - Improved error handling and retry logic for HTTP requests.
  - Refined batch sampling, data collation, and training step calculations for more robust training.
  - Simplified and unified KD loss computation by delegating to model-internal loss functions.
  - Extended plugin system with new hooks for training args and collators.

- **Bug Fixes**
  - Improved handling of incomplete batches and batch size calculations.
  - Fixed issues with plugin and collator selection logic.

- **Documentation**
  - Added detailed metadata and help descriptions for new training arguments.

- **Tests**
  - Updated and expanded end-to-end tests for knowledge distillation workflows.

- **Chores**
  - Refactored imports and type annotations for better maintainability and static analysis support.

### CODE REVIEW

File: src/axolotl/core/trainer_builder.py [hunk 1]
```diff
@@ -793,10 +793,6 @@ def build(self, total_num_steps):
             training_arguments_kwargs["kd_zscore_base_temp"] = (
                 self.cfg.kd_zscore_base_temp
             )
-        if self.cfg.kd_top_k_before_softmax is not None:
-            training_arguments_kwargs["kd_top_k_before_softmax"] = (
```
Comment (djsaunde) [inline src/axolotl/core/trainer_builder.py]: I assume doing top K before softmax messes with the logprobs?

File: src/axolotl/integrations/kd/collator.py [hunk 2]
```diff
@@ -47,6 +47,10 @@ class DataCollatorForKD(DataCollatorForSeq2Seq):
     position_pad_token_id: int = 0
     return_tensors: str = "pt"
 
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.tokenizer.deprecation_warnings["Asking-to-pad-a-fast-tokenizer"] = True
```
Comment (djsaunde) [inline src/axolotl/integrations/kd/collator.py:52]: What's this?

File: src/axolotl/integrations/kd/topk_logprob/forward_kl.py [hunk 3]
```diff
@@ -235,3 +211,76 @@ def topk_kd_loss_with_zscore(
         kd_loss = kd_loss / float(kd_loss_per_token.size(0))
 
     return kd_loss
+
+
+class ChunkedTopKKDLoss(nn.Module):
```
Comment (djsaunde) [inline src/axolotl/integrations/kd/topk_logprob/forward_kl.py:99]: 🚀

**Review by djsaunde — APPROVED** [review]
